File: jupyter/nb/source/pl09_memory_management/include/pl09.jl
import Printf

function main()
    argc = length(ARGS)
    s = if argc >= 1 parse(Int64, ARGS[1]) else 1000 * 1000 end
    m = if argc >= 2 parse(Int64, ARGS[2]) else 100 end
    n = if argc >= 3 parse(Int64, ARGS[3]) else m * 10 end
    if sizeof(Int64) * s * m > (1 << 30)
        Printf.@printf("you'd better not allocate that much memory")
        Printf.@printf("sizeof element(8) * s(%d) * m(%d) = %d > 1GB\n", s, m, sizeof(Int64) * s * m)
        exit(1)
    end
    a = Vector{Vector{Int64}}(undef, m)
    for i = 1:n
### if label == "measure_time"
        t0 = time_ns()
### endif
        b = fill(0, s)
### if label in [ "show_addr", "measure_time" ]
        p = pointer(b)
### endif
        a[(i - 1) % m + 1] = b
### if label == "measure_time"
        t1 = time_ns()
### endif
### if label == "show_addr"
        Printf.@printf("%ld\t%ld\n", i, Int64(p))
### endif
### if label == "measure_time"
        Printf.@printf("%ld\t%ld\t%ld\n", i, Int64(p), t1 - t0)
### endif
    end
end

main()

File: jupyter/nb/source/pl09_memory_management/include/pl09.rs
/*** if label == "measure_time" */
fn time_ns(now : &std::time::Instant) -> i64 {
    now.elapsed().as_nanos() as i64
}
/*** endif */
fn main() {
    let args : Vec<String> = std::env::args().collect();
    let argc = args.len();
    let s = if argc > 1 { args[1].parse::<usize>().unwrap() } else { 1_000_000 };
    let m = if argc > 2 { args[2].parse::<usize>().unwrap() } else { 100 };
    let n = if argc > 3 { args[3].parse::<usize>().unwrap() } else { m * 10 };
    let sizeof_elem = 8;
    if sizeof_elem * s * m > (1 << 30) {
        println!("you'd better not allocate that much memory");
        println!("sizeof element(8) * s({}) * m({}) = {} > 1GB", s, m, sizeof_elem * s * m);
        std::process::exit(1)
    }
    let b = Box::new(vec![0i64; s]);
    let mut a = Box::new(vec![b; m]);
/*** if label == "measure_time" */
    let ts = std::time::Instant::now();
/*** endif */
    for i in 0..n {
/*** if label == "measure_time" */
        let t0 = time_ns(&ts);
/*** endif */
        let b = Box::new(vec![0i64; s]);
/*** if label == "show_addr" or label == "measure_time" */
        let p = b.as_ptr() as usize;
/*** endif */
        a[i % m] = b;
/*** if label == "measure_time" */
        let t1 = time_ns(&ts);
/*** endif */
/*** if label == "show_addr" */
        println!("{}\t{}", i, p);
/*** endif */
/*** if label == "measure_time" */
        println!("{}\t{}\t{}", i, p, t1 - t0);
/*** endif */
    }
}
